TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+
+    NSMutableArray *dependencyNodes = [[NSMutableArray alloc] init];
+    [dependencies enumerateKeysAndObjectsUsingBlock:^(NSString *key, id object, BOOL *stop) {
+        [self pushPathComponent:key];
+
+        [self checkObject:key isOfClass:[NSString class] allowNil:NO];
+
+        if ([object isKindOfClass:[NSDictionary class]]) {
+            TWTJSONSchemaDepedencyASTNode *node = [[TWTJSONSchemaDepedencyASTNode alloc] initWithKey:key valueSchema:[self parseSchema:object]];
+            [dependencyNodes addObject:node];
+        } else {
+            NSArray *propertySet = [self parseArrayOfUnqiueStringsFromSchema:dependencies forKey:key];
+            TWTJSONSchemaDepedencyASTNode *node = [[TWTJSONSchemaDepedencyASTNode alloc] init];
+            node.key = key;
+            node.propertySet = propertySet;
+            [dependencyNodes addObject:node];
+        }
+
+        [self popPathComponent];
+
+    }];
+
+    [self popPathComponent];
+    return dependencyNodes;
+}
+
+
+// Keyword: items, not
+// Example: { "items" : {} }
+- (TWTJSONSchemaASTNode *)parseSchemaFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    NSDictionary *nestedSchema = schema[key];
+    if (!nestedSchema) {
+        return nil;
+    }
+
+    [self pushPathComponent:key];
+    TWTJSONSchemaASTNode *node = [self parseSchema:nestedSchema];
+    [self popPathComponent];
+    return node;
+}
+
+
+#pragma mark - Helper methods
+
+- (void)checkObject:(id)object isOfClass:(Class)className allowNil:(BOOL)allowNil
+{
+
+    if (allowNil && !object) {
+        return;
+    }
+
+    if ([object isKindOfClass:className]) {
+        return;
+    }
+
+    NSString *reason = [NSString stringWithFormat:@"Invalid type encountered. Expected object of class %@.", NSStringFromClass(className)];
+    [self failWithErrorCode:TWTJSONSchemaParserErrorCodeInvalidClass reason:reason object:object];
+}
+
+
+- (void)checkObjectIsArrayWithAtLeastOneItem:(id)object allowNil:(BOOL)allowNil
+{
+    [self checkObject:object isOfClass:[NSArray class] allowNil:allowNil];
+    if (!object || [object count] >= 1) {
+        return;
+    }
+
+    NSString *reason = @"Expected array with at least one item.";
+    [self failWithErrorCode:TWTJSONSchemaParserErrorCodeRequiresAtLeastOneItem reason:reason object:object];
+}
+
+
+- (void)checkObject:(id)object isOneOf:(NSSet *)validValues
+{
+    if ([validValues containsObject:object]) {
+        return;
+    }
+
+    NSString *reason = [NSString stringWithFormat:@"Invalid valid encountered. Expected one of %@", validValues];
+    [self failWithErrorCode:TWTJSONSchemaParserErrorCodeInvalidValue reason:reason object:object];
+}
+
+
+- (void)failWithErrorCode:(NSUInteger)code reason:(NSString *)reason object:(id)object
+{
+    NSString *pathString = [@"/" stringByAppendingString:[self.pathStack componentsJoinedByString:@"/"]];
+    object = object ?: [NSNull null];
+    reason = reason ?: @"";
```

Comment:
Don’t use this `?:` syntax. It’s a non-standard extension and not part of the C or Objective-C languages.